File: android/app/src/main/java/com/alexivanov/snapsell/handoff/ClipboardStager.kt
package com.alexivanov.snapsell.handoff

import android.content.ClipData
import android.content.ClipboardManager
import android.content.Context
import com.alexivanov.snapsell.domain.Money

/**
 * Puts the listing text on the clipboard as one block so the user can paste
 * it into the Marketplace form. Facebook has no public listing API; the
 * clipboard + album + deep link trio is the hand-off.
 */
object ClipboardStager {
    const val LABEL = "SnapSell listing"

    /**
     * Exact block format, so pasting into the title field first and then
     * cutting the rest into the description is predictable:
     *
     *     <title>
     *
     *     $<price>
     *
     *     <description>
     */
    fun format(title: String, price: Double, description: String): String =
        buildString {
            append(title.trim())
            append("\n\n")
            append(Money.usd(price))
            append("\n\n")
            append(description.trim())
        }

    fun copy(context: Context, title: String, price: Double, description: String) {
        val clipboard = context.getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager
        clipboard.setPrimaryClip(ClipData.newPlainText(LABEL, format(title, price, description)))
    }
}

File: ios/SnapSell/Handoff/ClipboardStager.swift
import UIKit

/// Stages listing text on the clipboard so it can be pasted into Marketplace.
/// The format (title, blank line, price, blank line, description) is shared
/// with Android; see the tests.
enum ClipboardStager {
    static func format(title: String, price: Double, description: String) -> String {
        let t = title.trimmingCharacters(in: .whitespacesAndNewlines)
        let d = description.trimmingCharacters(in: .whitespacesAndNewlines)
        return "\(t)\n\n\(Money.usd(price))\n\n\(d)"
    }

    static func stage(title: String, price: Double, description: String) {
        UIPasteboard.general.string = format(title: title, price: price, description: description)
    }
}
